### ISSUE TITLE
Lazy semantics are a mismatch with JavaScript

### ISSUE THREAD

**Issue by domenic** [zenparsing/es-observable#11]
I understand they fit better with C#, but I am not comfortable with how they deviate from JavaScript's iterator and promise designs.

This seems easily fixable, however. Simply execute the executor immediately. To get back laziness, simply use an observable-returning function, as is done today with promise and iterators.

**Comment by zenparsing** [comment]
I think laziness is actually necessary, given the other design choices.

Since Observable does not presuppose any kind of buffering, it can't begin pushing data until an observer has been subscribed.  Likewise, a combinator like "map" can't subscribe to its source until it has its own observer.  By induction we end up with laziness.  We have to construct the whole observable chain in advance.

What do you think?  Is that sound?

But I suppose you larger point remains:  can JS have lazy push-sequences and still remain consistent?

To be clear, I don't think that `async function*` should vend Observables, because of the reason you mention:  the laziness of Observables don't mesh well with the non-laziness of generators.

**Comment by blesh** [comment]
> This seems easily fixable, however. Simply execute the executor immediately. To get back laziness, simply use an observable-returning function, as is done today with promise and iterators.

@domenic This would break one of the many things that makes Observable a better async primitive than Promises, the ability to be retried or replayed wherever you have a reference.

> I don't think that async function* should vend Observables, because of the reason you mention: the laziness of Observables don't mesh well with the non-laziness of generators.

@zenparsing I'd argue that `function*` only vends something "lazy". If you never call `next` on the generator returned, it doesn't do anything. Given than it makes sense that the return of `function* async` would also be lazy.

```js
function* test() {
  console.log('started');
  yield 1;
}

var g = test();
// nothing has happened yet...

g.next();
// console logs "started"
```

So I think the entire "deviates from JavaScript's iterator and promise designs" is a non-issue. It's likely a good thing that it deviates from promise and iterator.

**Comment by domenic** [comment]
@blesh that is false (and your combative tone is not appreciated). Just get a reference to the observable-returning function. They are isomorphic.

@zenparsing still digesting your comment. However, it doesn't seem true that you can't push data without a subscription---it simply means the data doesn't go anywhere. Maybe this is the good-old hot-vs.-cold observables distinction.

**Comment by zenparsing** [comment]
@blesh I want to make it clear that we are in no way trying to establish Observable as a better async primative than promises.  Because observers can react to "return" and "throw", it means that you can, in principle, use observables instead of promises to model asynchronous return values.  But observables are really geared toward asynchronous *sequences*.

Regarding `function *`, what I mean is that calling a generator function will actually start the computation and then immediately pause it.  They are eagerly started, but lazily iterated over.  Observables, on the other hand, are lazily started and then pump data on their own until told to stop.

IMO, `async function *` should return *async iterators*, which can be (almost) trivially converted to an observable.  Here's my idea for async iterators (the repo is a little out of date, at the moment):

https://github.com/zenparsing/async-iteration

**Comment by domenic** [comment]
@jhusain's talk at JSConf actually clarified the asymmetry for me. He showed this:

```
interface Iterable {
  Generator [Symbol.iterator](void);
}

interface Observable {
  void [Symbol.observer](Generator);
}
```

The current proposal is instead `observable[Symbol.observer].subscribe(generator)`, if I am reading correctly. It makes a lot more sense to me to do `observable[Symbol.observer](generator)`. Does this clarify what I am talking about when I appeal to consistency?

**Comment by blesh** [comment]
@domenic My apologies if you found it combative. What part, specifically, was it that offended you? I just re-read it and I don't see anything I think could be misconstrued as combative.

Anyhow, it's not really *false*. It's an *advantage* of observable that anywhere you have a reference to one, you have the ability to retry or replay it, without having to keep a reference to whatever function and arguments made it. 

> I want to make it clear that we are in no way trying to establish Observable as a better async primative than promises. 

@zenparsing  Oh I know... **I think** they're a better async primitive. I never said that this repo was trying to establish that.

> Regarding function *, what I mean is that calling a generator function will actually start the computation and then immediately pause it. They are eagerly started, but lazily iterated over.

The code before the first `yield` isn't executed until you call `.next()` on the returned generator. So nothing at all in the `function*` is executed until that `.next()` is called on it's returned value. Unless the transpilers I've used have the implementations wrong? (totally possible I guess)

>  It makes a lot more sense to me to do observable[Symbol.observer](generator). 

@domenic I agree, although I the method name `iterator` and `observer` makes me cringe a little, I wish it was `iterate` and `observe`. RxJS 3, right now, is using `observer(generator)`

**Comment by zenparsing** [comment]
@domenic Sorry, I think some previous iterations might be confusing things, and I've been making late changes (like just now).  The current version aligns with Jafar's interface descriptions, except the return value is a cancellation function.

```js
interface Iterable {
  Generator [Symbol.iterator](void);
}

interface Observable {
  CancellationFunction [Symbol.observer](Generator);
}
```

Does that help?  Do you still see asymmetry in that light?

**Comment by domenic** [comment]
It does help clarify, by bringing my confusion into clearer focus :). I can now more clearly state what troubles me:

Right now it seems like the `Observable` class is just a package for a function, the executor. That is, instead of `doStuff(sink)`, there's an object with a method you call in order to do stuff to the sink: `observable[@@observer](sink)`. This essentially calls `executor(sink)` (with some indirection to ensure invariants).

This seems very kingdom-of-nouns-ey. We don't need such "delegate interfaces" in JavaScript---the difference between `observable[@@observer](sink)` and `observable(sink)` seems thin.

In other words, why is it

```js
var x = new Observable(executor);
x[@@observer](sink);
x[@@observer](sink2);
```

instead of

```js
var f = createObservable(executor);
f(sink);
f(sink2);

// here createObservable wraps executor to ensure invariants
```

If the observable were eager, then things would make more sense to me. Instead of just being a package for indirectly calling `executor`, it would represent an ongoing operation that was already underway. The executor would have already executed, and calling `.subscribe` would represent someone wanting to start listening to that operation, which currently has no listeners.

**Comment by zenparsing** [comment]
Gotcha.  So you're saying that Observable might just as well be a function, since all it does is wrap a function (and potentially provides some nice methods).

**Comment by zenparsing** [comment]
So the advantage of having an observable object that wraps the executor is the associated methods.

Let's say that we have `async for of`, which gets an async iterator using `Symbol.asyncIterator`.  And let's say that we have a "listen" function that creates an observable for DOM events.  With the appropriate method on Observable.prototype, we can join the two:

```js
async for (let event of listen(myElement, "mousedown")) {
    // Do something with event
}
```

I suppose that a function instance could also have a `Symbol.asyncIterator`, but that seems odd?

**Comment by domenic** [comment]
Yeah, it seems to come back to the old functions-versus-methods issue (similar to the bind operator's use cases). That is, Observable as formulated is just a place to put methods, so you can do

```js
getObservable().map(a).filter(b).forEach(c);
```

instead of

```js
forEach(filter(map(getObservable(), a), b), c)
```

or

```js
getObservable()::map(a)::filter(b)::forEach(c);
```

The fact that `async for of` would not work with the bind operator is reminiscent of our many conversations about how it does not play well with polymorphic dispatch.

It's a believable idea, to create such wrapper objects for specific types of functions. It would be good to modify the pedagogy, e.g. "observables are wrappers for functions that operate on generator objects" instead of "observables represent an async sequence via their executor function" (which has a false analogy to promise executor functions and promises representing single values). Maybe doing so would resolve the OP issue, although I think it might be worth re-thinking areas in which this proposal attempts to align with promises and iterables before doing so. (Executor naming is off, of course, but also such things as the symbol---what protocol is it intended to support? Etc.)

But yeah, I'm not sure introducing a whole new built-in to conceptualize "functions that operate on generator objects" and give them some helper functions written left-to-right is that compelling, especially given the bind operator. E.g. it's not any clearly more compelling than "functions that operate on iterable objects" or "functions that operate on promise objects" or "functions that operate on Map objects."

**Comment by jhusain** [comment]
Agreed. Observable is only a class to enable method chaining via prototypes.

**Comment by domenic** [comment]
A better analogy than the ones in my last sentence: it's not clearly more important to have an Observable class (as a wrapper around functions that call arg.next(v) multiple times, then one of arg.throw(e) and arg.return(v)) than it is to have a Callback class as a wrapper around functions that call arg(e) or arg(null, v) once. That would be the path taken by "thunk" advocates in Node.js (i.e. functions that return callbacks). Instead we added promises, which are quite different.

**Comment by jhusain** [comment]
So that we can talk about the benefits and drawbacks of this pattern let's give it a name: **Function Class** (if it already has a name let me know). A Function Class is a class instance that contains a single function member. Function classes solve the following problem: While combinators are just as applicable to functions as they are objects, it is not ergonomic to use  function combinators in JS because functions lack prototypes. Function classes enable ergonomic, polymorphic combinator chaining over function.

@domenic Was the decision to use Promises partially motivated by the desire to avoid introducing a Function class? Or were you merely observing that there is no real precedent here?

Even if we had bind there would still be a good reason to create an Observable Function class: there are  _demonstrated polymorphic use cases_ for Observables. Sorry I forgot about these use cases earlier.

In C# it is common for DAO to expose  subclasses of Observable that override some of the combinators. Whenever an overridden combinator is called on the DerivedObservable (map, filter, etc) it is added to a list of operations to be performed. When the DerivedObservable's subscribe method is called, the operations are remoted to the data source (ex. Service, or database) instead of being executed locally. This allows data transformations to be performed closer to the data. If a combinator inherited from the Observable base class is chained after a remoted combinator it is executed locally on the result of the remotely transformed stream. 

```JavaScript 
twitterAPI.
   // TwitterObservable
   tweets.
     //overridden to use parse function.toString and translate recognized function expression patterns into query string parms (&contains=ES2016). 
     // in the event pattern can not be recognized, we delegate to the base class filter implementation
     filter(tweet => tweet.body.contains("ES2016")).
     // ditto (&mazSize=10)
     take(10).
     // flatMap inherited from Observable and executed locally
     flatMap(tweet => translate(tweet.body, "fr-CA")).
     subscribe(::console.log);
```

Another similar case for polymorphism would be reducers/scanners. FYI scan is function just like a reduce except returns all intermediary computations, not just the final value:

```JavaScript
[1,2,3].((acc, curr) => acc + curr) // [1, 3, 6]
```
Multiple chained combinators that can be expressed as scans (map, filter, take, etc) can be collapsed into a single scan operation. This can significantly improve performance.

To transparently achieve performance gains observable combinators could emit a ScannableObservable subclass when any combinator which could be expressed in terms of scan was invoked. A ScannableObservable would override all operations that could be efficiently expressed as a scan (ex. map, filter). Whenever such an operation was called the ScannableObservable would simply be add the combinator's aggregation function to a list. When subscribe was called on a ScannableObservable, all its accumulated aggregation functions would be composed into one aggregation function, allowing that subset of the stream transformation to be executed with a single scan operation. As soon as a combinator method was called which could not be expressed efficiently as a scan (zip, flatMap), the default Observable implementation would be used. 

```JavaScript
observable.
   // returns ScannableObservable
   map(x => x +1).
   // returns another ScannableObservable
   filter(x => x > 5).
   // returns an Observable
   flatMap(x => Observable.of(x + 10)). 
   subscribe(::console.log);
```

In the code above the map and filter operations are executed with one scan operation, whereas flatMap has the same cost as it normally does.

To summarize polymorphic Observable composition has been used to great effect on other platforms. This is an argument in favor of an Observable function class (and probably function classes in general) regardless of the status of the bind operator.

**Comment by domenic** [comment]
> Function Class

Yes, this is great. I am glad we are having the discussion on this level. For example, this opens up questions like: should Function Classes be subclasses of Function? Positioning observables as function classes makes much more sense to me than positioning them as in some way analogous to iterables.

> Was the decision to use Promises partially motivated by the desire to avoid introducing a Function class? Or were you merely observing that there is no real precedent here?

The latter. Plus the fact that promises have more precedent and provide more than "thunks" (or their function class equivalent) do, similar to how async iterable over observable.

>  there are demonstrated polymorphic use cases for Observables. Sorry I forgot about these use cases earlier.

This is also great stuff. However, it raises the burden on this repo for proving its worth. It means that for this proposal to be worthwhile, not only do we need to see the ecosystem adopt observables of this sort, but also need to see the ecosystem subclassing observables and doing this kind of polymorphic work.

> This is an argument in favor of an Observable function class (and probably function classes in general) regardless of the status of the bind operator.

I'm interested in whether we can extract a more general pattern here. In general the idea of a function class seems to  be specializing methods for functions which have a particular signature. In this case the signature is `Generator -> CancelationFunction`, but there are many other cases where this could be useful. For example, a function class built around `Promise -> void` (or `CancelablePromise -> CancelationFunction`) allows lazy transformations on single asynchronous values. `Iterable -> void` is reminiscent of some recent TC39 discussions about how to add generic-iterable methods. (You might think this is subsumed into a generator object-accepting function class, but I don't think so: it can be useful to have the invariant that you only assume a .next(), and not a .return() or .throw().) Arguably `SelectorString -> DOMElementArray` is its own interesting class.

**Comment by benjamingr** [comment]
Very interesting points here @domenic. 

Sorry to backtrack a little and it's probably obvious, but was adding map/reduce/filter etc ever considered for iterables? The reason laziness was added to C# was because LINQ requires it (it compiles expressions you input through adapters to other formats like SQL) but @zenparsing raises some interesting points. So, I agree that laziness fits better with the C# ecosystem at the moment - there _are_ lazy projects in JS (lazyjs with over 2000 stars, Facebook's ImmutableJS and so on) but they're not nearly as common. For one to be fair we have to admit that laziness is usually harder than strict evaluation, especially to people who are not already familiar with it.

> I'm interested in whether we can extract a more general pattern here. In general the idea of a function class seems to be specializing methods for functions which have a particular signature. 

Well, I think a good example would be looking at something like Ramda vs something like lodash (or even lodash-fp vs loadsh). 

@jhusain 

> So that we can talk about the benefits and drawbacks of this pattern let's give it a name: Function Class

I'm wondering - in abstract references' (now bind operator) original proposal there was a way to do this sort of polymorphic dispatch. I think it was called protocols and I wonder what happened to it - perhaps @zenparsing could shed some light. I'm not sure where this is standing today - here is a gist https://gist.github.com/genericallyloud/7086380

**Comment by rpominov** [comment]
Just to add 2 cents: observables indeed very similar to functions. We can define many methods that operate on functions similar to methods that operates on observables:

```js
function map(fn, cb) {
  return (x) => {
    cb(fn(x))
  }
}

function filter(fn, cb) {
  return (x) => {
    if (fn(x)) {
      cb(x)
    }
  }
}
```

But what is important, is that observables invert dependency. With callbacks we apply `map/filter/etc` directly to callback, so we need the callback at the beginning. While with observables we can apply transformations and only then inject the callback (subscribe). This inversion is very important if we want define `flatMap`, which is impossible to define as an operation on callbacks.

So observables not only containers for methods, but they also do the inversion of dependency, which is important.

Perhaps all this is obvious, but still wanted to share these thoughts, i think they are relevant to the discussion.

**Comment by domenic** [comment]
> But what is important, is that observables invert dependency. With callbacks we apply map/filter/etc directly to callback, so we need the callback at the beginning. While with observables we can apply transformations and only then inject the callback (subscribe).

This is an important point to highlight. However, it's not specific to observables. It works with any lazy function-class. For example, here is a function-class for lazy operations on promises. It allows you to apply the transformations and then later inject the promise (subscribe):

```js
class Promiseable {
  constructor(operationOnPromise) {
    this._operationOnPromise = operationOnPromise;
  }
  
  subscribe(promise) {
    this._operationOnPromise(promise);
  }
  
  if(predicate) {
    return new this.constructor[Symbol.species](p => {
      if (predicate()) {
        this._operationOnPromise(p);
      }
    });
  }
  
  static get [Symbol.species] { return this; }
}

// Usage
const logger = new Promiseable(p => {
  p.then(v => console.log(v), e => console.error(e));
});

logger.subscribe(Promise.resolve(10)); // will log 10

const sometimesLogger = logger.if(() => Math.random() > 0.5);

// Each of these has a 1/2 chance of actually logging
sometimesLogger.subscribe(Promise.resolve(1));
sometimesLogger.subscribe(Promise.resolve(2));
sometimesLogger.subscribe(Promise.reject(new Error("boo")));
```

**Comment by rpominov** [comment]
Sorry, maybe there is a bit of misunderstanding. Originally my point on inversion of dependency was against this: 

> This seems very kingdom-of-nouns-ey. We don't need such "delegate interfaces" in JavaScript---the difference between `observable[@@observer](sink)` and `observable(sink)` seems thin.

(did you meant `executor(sink)`, btw?)

Promises, btw, also invert the dependency — that why we use them instead of callbacks :)

**Comment by domenic** [comment]
Ah yes, my thoughts have evolved a decent bit since the OP (this thread is pretty misnamed by now, but I don't want to rename it because that breaks mail-client threading... personal phobia). I understand the importance of function-classes now but think this proposal should be more focused around that general idea instead of a specific type of function-class, i.e. functions-taking-generators.

**Comment by jhusain** [comment]
I understand your point about wanting to explore the idea of function classes in general. I agree it's a useful concept to articulate.

The question of whether Function classes should subclass Function are pretty interesting. The practical problem I see is that it can't be polyfilled (correct me if I'm wrong) which seems serious. I'm also not sure it's more ergonomic. Notably .NET chose the kingdom of nouns approach for this very reason despite having the ability to add methods to function types (delegate + extension methods). Doesn't mean this is the right choice for JS, just a datapoint.

JH

**Comment by domenic** [comment]
Functions can definitely be subclassed in ES2015. In ES5 you'd need to use other tricks (factory function with instance methods copied from the prototype, basically) and it wouldn't be complete. But assuming ES2015 as a baseline for polyfills seems reasonable.

**Comment by benjamingr** [comment]
This discussion goes over my head - function classes sound very interesting to support but... would someone mind explaining to me how function classes would obsolete observable or how it would be a function class anyway? 

I mean, it's definitely interesting but I'm not sure how it's directly related to observables.

**Comment by domenic** [comment]
Actually subclassing is fine with full fidelity in ES5 by using return-override.

@benjaminr observables are a type of function-class. Compare the polyfill with my Promiseable to see he commonality.

**Comment by benjamingr** [comment]
@domenic thanks, contrasting the two did the trick. It reminds me a lot of the way Erik Meijer defines iterators/observables in [duality and the end of reactive](http://channel9.msdn.com/Events/Lang-NEXT/Lang-NEXT-2014/Keynote-Duality) as getters/setters. Very interesting.

**Comment by zenparsing** [comment]
Closing this for cleanup.
